### PR TITLE
correct the codegen of floating point subtypes

### LINKED ISSUE #21216: reinterpret segfault
The following behavior only seems to exist on linux current master (1dd10aa). Discovered in the DecFP.jl package (current master), travis logs [here](https://travis-ci.org/stevengj/DecFP.jl/jobs/216194537). Repro:

```julia
julia> using DecFP

julia> Base.bitcast(UInt128, Dec128(1.5))
0x303e000000000000000000000000000f

julia> reinterpret(UInt128, Dec128(1.5))

signal (11): Segmentation fault
while loading no file, in expression starting on line 0
reinterpret at ./essentials.jl:0
unknown function (ip: 0x7f71512b8a25)
jl_call_fptr_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:326 [inlined]
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:345 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball64/build/src/gf.c:2234
do_call at /home/centos/buildbot/slave/package_tarball64/build/src/interpreter.c:75
eval at /home/centos/buildbot/slave/package_tarball64/build/src/interpreter.c:242
jl_interpret_toplevel_expr at /home/centos/buildbot/slave/package_tarball64/build/src/interpreter.c:34
jl_toplevel_eval_flex at /home/centos/buildbot/slave/package_tarball64/build/src/toplevel.c:577
jl_toplevel_eval_in at /home/centos/buildbot/slave/package_tarball64/build/src/builtins.c:484
eval at ./boot.jl:235
unknown function (ip: 0x7f736cdb766f)
jl_call_fptr_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:326 [inlined]
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:345 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball64/build/src/gf.c:2234
eval_user_input at ./REPL.jl:66
unknown function (ip: 0x7f7151278b76)
jl_call_fptr_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:326 [inlined]
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:345 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball64/build/src/gf.c:2234
macro expansion at ./REPL.jl:97 [inlined]
#1 at ./event.jl:73
unknown function (ip: 0x7f7151275caf)
jl_call_fptr_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:326 [inlined]
jl_call_method_internal at /home/centos/buildbot/slave/package_tarball64/build/src/julia_internal.h:345 [inlined]
jl_apply_generic at /home/centos/buildbot/slave/package_tarball64/build/src/gf.c:2234
jl_apply at /home/centos/buildbot/slave/package_tarball64/build/src/julia.h:1416 [inlined]
start_task at /home/centos/buildbot/slave/package_tarball64/build/src/task.c:261
unknown function (ip: 0xffffffffffffffff)
Allocations: 3683352 (Pool: 3681747; Big: 1605); GC: 5
Segmentation fault
```

As far as I can tell, `reinterpret(UInt128, x)` just calls `Base.bitcast`, so I'm not sure why one segfaults and the other seems to work just fine.

### PR DESCRIPTION
**Pull request by vtjnash**
avoids casting away floating point bits until used in a floating-point operation
also ensures that alignment is consistent in the LLVM struct representation as in the Julia struct representation computation

fix #21216

### CODE REVIEW

File: src/cgutils.cpp [hunk 1]
```diff
@@ -402,26 +402,17 @@ static Type *bitstype_to_llvm(jl_value_t *bt)
         return T_int8;
     if (bt == (jl_value_t*)jl_long_type)
         return T_size;
+    if (bt == (jl_value_t*)jl_float32_type)
```
Comment (vchuravy) [inline src/cgutils.cpp]: Can we add support for Float16 here? (Deactivated right now?)

Comment (vtjnash) [inline src/cgutils.cpp]: This isn't supporting Float16, this is defining the C ABI conversion layer (which usually either doesn't define `half float`, or defines that it is implicitly converted to `float`). This commit shouldn't have any effect on the compiler float16 intrinsic support.

**Comment by tkelman** [comment]
backporting this to release-0.5 breaks the ARM build, unfortunately

**Comment by StefanKarpinski** [comment]
Is there any urgent need to have this in 0.5.2?

**Comment by vtjnash** [comment]
`jl_subtype` needs a third argument (a `0`) on v0.5

**Pull request merged**